src/include/txn.i

```diff
@@ -300,42 +348,49 @@ __wt_txn_modify(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd)
 		    "Attempt to update in a read-only transaction");
 
 	WT_RET(__txn_next_op(session, &op));
-	op->type = F_ISSET(session, WT_SESSION_LOGGING_INMEM) ?
-	    WT_TXN_OP_INMEM_COL : WT_TXN_OP_BASIC_COL;
-#ifdef HAVE_TIMESTAMPS
-	if (__wt_txn_update_needs_timestamp(session, op)) {
-		__wt_timestamp_set(&upd->timestamp, &txn->commit_timestamp);
+        if (F_ISSET(session, WT_SESSION_LOGGING_INMEM)) {
+                if (btree->type == BTREE_ROW)
+                        op->type = WT_TXN_OP_INMEM_ROW;
+                else
+                        op->type = WT_TXN_OP_INMEM_COL;
+        } else {
+                if (btree->type == BTREE_ROW)
+                        op->type = WT_TXN_OP_BASIC_ROW;
+                else
+                        op->type = WT_TXN_OP_BASIC_COL;
+        }
+	op->u.op_upd = upd;
+	upd->txnid = session->txn.id;
 
-		/*
-		 * Checkpoint transactions and lookaside transactions will not
-		 * be prepared.
-		 */
-		if (!WT_SESSION_IS_CHECKPOINT(session) &&
-		    !F_ISSET(btree, WT_BTREE_LOOKASIDE) &&
-		    !WT_IS_METADATA(op->btree->dhandle)) {
-			/*
-			 * Store the key, to search the prepared update in case
-			 * of prepared transaction.
-			 */
-			if (btree->type == BTREE_ROW) {
-				WT_ITEM key;
-				WT_RET(__wt_cursor_get_raw_key(&cbt->iface,
-				    &key));
-				WT_RET(__wt_buf_set(session, &op->u.op_row.key,
-				    key.data, key.size));
-				op->type = F_ISSET(session,
-				    WT_SESSION_LOGGING_INMEM) ?
-				    WT_TXN_OP_INMEM_ROW : WT_TXN_OP_BASIC_ROW;
-			} else
-				op->u.op_col.recno = cbt->recno;
-		}
-	}
+#ifdef HAVE_TIMESTAMPS
+	__wt_txn_update_set_timestamp(session, op, false, NULL);
+        /*
+         * Copy the key into the transaction op structure, so the update
+         * can be evicted to lookaside, and we have a chance of finding it
+         * again. This is only possible for transactions that are in the
+         * prepared state, but we don't know at this stage if a transaction
+         * will be prepared or not.
+         */
+        if (!WT_SESSION_IS_CHECKPOINT(session) &&
```

Comment:
@bvpvamsikrishna this is quite a significant change - I moved some functionality inside `__wt_txn_update_needs_timestamp` (and renamed it). When I was resolving merge conflicts, I got to thinking that this code wasn't working as expected before. I believe it always needs to save a copy of the key here - not only when the transaction already has a timestamp set.
That could explain why we didn't see any performance change with the original merge of WT-4246.

Reply:
You observation is correct, there is some problem here. But, it is other way around for optimisation reason. 
This is required for prepared transactions and an update done by a transaction, which will go to prepared state will not have a timestamp. So, this code to be enabled only when the update doesn't have a timestamp.